src/gpu/generic/sycl/ref_inner_product.hpp

```diff
@@ -0,0 +1,175 @@
+            auto scratchpad = scratchpad_registry().registrar();
+            scratchpad.book(memory_tracking::names::key_nested,
+                    matmul_pd->scratchpad_registry());
+            return status::success;
+        }
+
+        std::shared_ptr<primitive_desc_t> matmul_pd;
+
+    private:
+        bool check_if_dtypes_valid(const data_type_t &src_dt,
+                const data_type_t &dst_dt, const data_type_t &bias_dt,
+                const data_type_t &weight_dt) const {
+            using namespace data_type;
+            return (utils::one_of(src_dt, f32) && utils::one_of(weight_dt, f32)
+                           && utils::one_of(dst_dt, f32)
+                           && utils::one_of(bias_dt, f32, undef))
+                    || (utils::one_of(src_dt, f16)
+                            && utils::one_of(weight_dt, f16)
+                            && utils::one_of(dst_dt, f16, f32, s8, u8)
+                            && utils::one_of(bias_dt, f16, f32, undef))
+                    || (utils::one_of(src_dt, u8, s8)
+                            && utils::one_of(weight_dt, s8)
+                            && utils::one_of(dst_dt, u8, s8, s32, bf16, f32)
+                            && utils::one_of(
+                                    bias_dt, u8, s8, s32, bf16, f32, undef))
+                    || (utils::one_of(src_dt, bf16)
+                            && utils::one_of(weight_dt, bf16)
+                            && utils::one_of(dst_dt, f32, bf16)
+                            && utils::one_of(bias_dt, f32, bf16, undef));
+        }
+
+        std::vector<int> get_dim_order(int ndims, const dims_t strides) {
+            std::vector<int> order(ndims);
+            for (int i = 0; i < ndims; ++i) {
+                order[i] = i;
+            }
+
+            std::sort(
+                    order.begin(), order.end(), [&strides](size_t i, size_t j) {
+                        return strides[i] < strides[j];
+                    });
+
+            return order;
+        }
+
+        status_t create_ip_mds() {
+            auto accumulate_dimensions = [](const dims_t dimensions, int start,
+                                                 int end) -> int64_t {
+                int64_t accum = 1;
+                for (int i = start; i < end; i++) {
+                    accum *= dimensions[i];
+                }
+                return accum;
+            };
+
+            const auto src_md_ = arg_md(DNNL_ARG_SRC);
+            const auto weights_md_ = arg_md(DNNL_ARG_WEIGHTS);
+            const auto bias_md_ = arg_md(DNNL_ARG_BIAS);
+            auto src_wrap = memory_desc_wrapper(src_md_);
+            auto w_wrap = memory_desc_wrapper(weights_md_);
+
+            // src and weights dims need to be in the same order
+            if (get_dim_order(src_wrap.ndims(), src_wrap.strides())
+                    != get_dim_order(w_wrap.ndims(), w_wrap.strides())) {
+                return status::unimplemented;
+            }
```

Comment:
Highlighting this as we are currently skipping the cases where src and weights are in a different order. We can consider supporting this in the future using reorder.